docs/DESIGN.md

```diff
@@ -0,0 +1,473 @@
+be able to differentiate temporary errors (backoff then retry) from permanent
+errors (give up).
+
+### Implementation detail: closing forwarded connections
+
+Using the go standard library, forwarding can be implemented as a pair of
+single-direction copy operations, each between the same pair of `net.Conn`
+connections.
+
+Consider a single copy operation from a src conn to a dst conn. Assume
+the copy operation has a timeout set and may also complete after copying a
+fixed number of bytes. There are a variety of reasons a copy might complete,
+and some require the server to take different actions:
+
+```
+reason for completion       outcome      server action
+---------------------       -------      -------------------------
+buffer copied, no EOF       okay         reset timeout, continue another copy
+src conn EOF                done         shutdown writing dst conn, return OK
+dst read error              fail         ensure both conns closed, return error
+src write error             fail         ensure both conns closed, return error
+timeout, no bytes copied    fail         ensure both conns closed, return error
+timeout, 1+ bytes copied    okay         reset timeout, continue another copy
+```
+
+Ensuring writing is shutdown on the dst conn after seeing an EOF from the
+src conn is intended to send all data read from src before EOF to the dst, and
+then inform the dst that no more messages will be coming.  The `net.Conn`
+abstraction doesn't quite suffice, we also need `CloseWrite` to handle this
+scenario.
+
+```
+dst conn protocol   message sent to inform other party of EOF
+-----------------   -----------------------------------------
+TCP                 FIN
+TLS                 alert close_notify
+```
+
+Ref: "3.5. Closing a Connection" in [rfc793](https://www.ietf.org/rfc/rfc793.txt).
+
+Complication: the "bytes copied" indicator used to express the application
+idle timeout needs to consider bytes copied in both directions, from this
+single copy operation and the anti-parallel one, otherwise it may 
+application connections where one direction of the conversation is silent, but
+the other is very busy. Maybe there's a cleaner way to express it.
+
+### TLS
+
+There is a tradeoff between security and supporting backwards compatibility 
+with clients that may only implement known-vulnerable protocols.
+
+In ascending order of security and decreasing compatibility, we have:
+
+1. server is willing to entertain TLS 1.0 connections, and a wide variety of 
+   cipher suites
+2. server accepts TLS 1.2 or TLS 1.3 only, but refuses to use particularly 
+   broken cipher suites
+3. server accepts TLS 1.3 only, limited to the
+   [set of cipher suites](https://datatracker.ietf.org/doc/html/rfc8446#section-9.1)
+   that the TLS 1.3 standard dictates TLS-compliant applications MUST or
+   SHOULD support
+4. custom protocol: server accepts connections with TLS 1.3 only, using a 
```

Comment:
another thought I had after writing this all up was that in some parts of the design the tradeoffs are made to support being used in a wide range of scenarios at the expense of being great at any niche (e.g. supporting arbitrary application protocols) but in other parts of the design i propose a tradeoff in a wildly different direction.

from a product adoption perspective maybe it'd be better to pick either "jack of all trades LB - support broad range of scenarios" or "niche LB, pick a specific scenario and try to delight users for that scenario" then align all the design decisions to support that.

that suggests trying to dial back in the direction of supporting a wider range of situations here to match supporting a wide range of application protocols -- TLS1.3

i'm certainly no cryptographer / security engineer

Reply:
It can be difficult to figure out the purpose of challenge programs, given that they don't have an actual practical purpose (other than ultimately being a challenge/exercise).

I think it's alright to draw the line at "prototype-y, secure by default, jack-of-all-trades LB", which for me lands somewhere around layer4 + TLS1.3.

Reply:
I will leave this bit of the document as is, but will reword below to clarify we will implement option 3, TLS1.3  , and not go restrict things further